# Deriving a random subset of tweets (JSON or CSV)

Let's say you're working with a file containing tweets, and you'd like to derive a random sample of tweets from that file.  For example, your set might be in chronological order, and you want just a random sample of 100 tweets from a mix of dates.

Note that the technique we describe here works for any type of data file with one observation per line, whether it's tweets or some other type of data, and whether the observations are in CSV format, JSON format, or some other text-based format with one observation per line.

Let's take a look at our sample data file:

In [26]:
!head -n 5 debatetweets.csv

It appears to have a header row, which we'll deal with shortly, followed by the data.  It looks like the tweets were written to the file in some sort of chronological order.  How long is it?

In [27]:
!wc -l debatetweets.csv

    1001 debatetweets.csv


It's 1,001 rows, so that's 1 header row followed by 1,000 observations.

The critical tool for accomplishing this cleverly is the `shuf`/`gshuf` shell command.  This command, part of the GNU Coreutils library, shows up as `shuf` in a bash shell, or as `gshuf` in Mac OSX (installed via `brew install coreutils`) and as `gshuf` in the IPython shell like we have here.  Let's see what it can do:

In [28]:
!gshuf --help

Usage: gshuf [OPTION]... [FILE]
  or:  gshuf -e [OPTION]... [ARG]...
  or:  gshuf -i LO-HI [OPTION]...
Write a random permutation of the input lines to standard output.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
  -e, --echo                treat each ARG as an input line
  -i, --input-range=LO-HI   treat each number LO through HI as an input line
  -n, --head-count=COUNT    output at most COUNT lines
  -o, --output=FILE         write result to FILE instead of standard output
      --random-source=FILE  get random bytes from FILE
  -r, --repeat              output lines can be repeated
  -z, --zero-terminated     line delimiter is NUL, not newline
      --help     display this help and exit
      --version  output version information and exit

GNU coreutils online help: <http://www.gnu.org/software/coreutils/>
Full documentation at: <http://www.gnu.org/software/coreutils/shuf>
or available locally via:

It looks like this is something we can use!  First we need to peel off the header row; we'll put it back later:

In [29]:
!head -n 1 debatetweets.csv > header.csv

Now for the main event.  We'll pipe ( `|` ) everything *except* the first line of debatetweets.csv to `gshuf`, and we'll take advantage of the `-n` option to request only 100 lines:

In [30]:
!tail -n +2 debatetweets.csv | gshuf -n 100 > only100tweets.csv

Now let's quickly size up what we got out:

In [31]:
!head -n 5 only100tweets.csv

That looks like the random sample we expected.  It doesn't appear to be in any chronological order.

In [32]:
!wc -l only100tweets.csv

     100 only100tweets.csv


Last but not least, we do need to reattach the header row (not applicable in the case of a line-oriented JSON file):

In [33]:
!cat header.csv only100tweets.csv > debatetweets-100sample.csv

In [34]:
!head -n 5 debatetweets-100sample.csv

Done!